### PR TITLE
binlogplayer: Enable throttler in filtered replication.

### PR DESCRIPTION
**Pull request by michael-berlin**
@alainjobart

---
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/1707)

### CODE REVIEW

File: go/vt/tabletmanager/binlog_players.go [hunk 1]
```diff
@@ -94,7 +94,12 @@ type BinlogPlayerController struct {
 	lastError error
 }
 
+// newBinlogPlayerController instantiates a new BinlogPlayerController.
+// Use Start() and Stop() to start and stop it.
+// Once stopped, you should call Close() to stop and free resources e.g. the
+// healthcheck instance.
 func newBinlogPlayerController(ts topo.Server, vtClientFactory func() binlogplayer.VtClient, mysqld mysqlctl.MysqlDaemon, cell string, keyRange *topodatapb.KeyRange, sourceShard *topodatapb.Shard_SourceShard, dbName string) *BinlogPlayerController {
+	healthCheck := discovery.NewHealthCheck(*binlogplayer.BinlogPlayerConnTimeout, *healthcheckRetryDelay, *healthCheckTimeout, "" /* statsSuffix */)
 	blc := &BinlogPlayerController{
```
Comment (alainjobart) [inline go/vt/tabletmanager/binlog_players.go]: can just return here:
return discovery.NewHealthCheck(...)

**Comment by alainjobart** [comment]
LGTM

**Comment by enisoc** [comment]
:lgtm: I did a drive-by review because I saw binlogs mentioned :)

<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> LGTM
> 
> [Approved with PullApprove](https://pullapprove.com/youtube/vitess/pull-request/1707/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=youtube/vitess)
</details>


---

Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[go/vt/binlog/binlogplayer/binlog_player.go, line 396 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4jWXSDpEsXLHU-3UV:-KI4jWXTMg2uWdZvbJ5o:-2110690486) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/binlog/binlogplayer/binlog_player.go#L396)):*
> ```Go
> 				break
> 			}
> 			time.Sleep(backoff)
> ```

I'm guessing `backoff` is typically small enough that we don't care to break from the sleep if the context is cancelled?

---

*[go/vt/tabletmanager/binlog_players.go, line 217 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4lUD5a3ruj1dfPEWm:-KI4lUD6tPoXRYwFxZqu:666742396) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/tabletmanager/binlog_players.go#L217)):*
> ```Go
> 	if bpc.ctx != nil {
> 		bpc.playerMutex.Unlock()
> 		log.Fatalf("%v: cannot Close() a BinlogPlayerController which was not Stop()'d before", bpc)
> ```

Do we need to kill the process? Why not panic? Or even returning an error from `Close()` (which many `Close()` methods do in the stdlib)?

---

*[go/vt/throttler/throttler.go, line 35 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4mJoU7sww3bTZWbTV:-KI4mJoVqEsXmszwORRJ:1219238559) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/throttler/throttler.go#L35)):*
> ```Go
> 	// MaxRateModuleDisabled can be set in NewThrottler() to disable throttling
> 	// by a fixed rate.
> 	MaxRateModuleDisabled = int64(math.MaxInt64)
> ```

Why do you prefer a typed constant here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1707#-:-KI4mi3bbqFLW6ki8sbQ:-951246880)*

**Comment by michael-berlin** [comment]
Review status: 10 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[go/vt/binlog/binlogplayer/binlog_player.go, line 396 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4jWXSDpEsXLHU-3UV:-KI4pxeEI--KERPYQbwn:-1905916391) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/binlog/binlogplayer/binlog_player.go#L396)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> I'm guessing `backoff` is typically small enough that we don't care to break from the sleep if the context is cancelled?
</details>
The maximum backoff duration is 1 second. (I've clarified this in the comment of the Throttler.Throttle() method.)

Usually, it's 1 second / rate e.g. a rate of 1000 TPS will result into a backoff of 1 ms.

Given this, I would not add a select for a cancelled context. Let me know if you think otherwise.

---

*[go/vt/tabletmanager/binlog_players.go, line 103 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4KFznt1-3DrTF2-MI-r1-103:-KI4quAlr6hFvKCWO58k:-499146009) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/tabletmanager/binlog_players.go#L103)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> can just return here:
> 
> return &BinlogPlayerController {
</details>
Done.

---

*[go/vt/tabletmanager/binlog_players.go, line 217 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4lUD5a3ruj1dfPEWm:-KI4rIL4x0KguRuK9JbC:1848643163) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/tabletmanager/binlog_players.go#L217)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Do we need to kill the process? Why not panic? Or even returning an error from `Close()` (which many `Close()` methods do in the stdlib)?
</details>
My intention was to have something like an assertion here. I overlooked that log.Fatalf() does something else than panic(). Thanks for pointing this out :)

So panic() would make sense here. However, Iteration() does call recover() and catch any panics. Therefore, it basically renders any panic() into an error and we could use an error in the first place.

I was also thinking about returning an error but I wanted to have something stronger in the style of an assertion here because I consider it a programming error to call Close() on a non-stopped binlog player.

Returning an error instead of crashing the process would be better if we assume that our tests do not cover all programming errors. This way we avoid the worst case that a destination master in production could (constantly) crash. For empty (non-serving) destination shards this is not a problem. However, we also want to reshard into existing shards in the future. Given that I'm learning towards returning an error here.

The fact that there's already a panic handler also suggests that I shouldn't kill the process here.

Thoughts?

---

*[go/vt/throttler/throttler.go, line 35 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4mJoU7sww3bTZWbTV:-KI4uJFsPI8uZ0bP-7LL:-1209850382) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/throttler/throttler.go#L35)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Why do you prefer a typed constant here?
</details>
With typed constant you mean the explicit cast?

I had int64(-1) in a previous version because the compiler complained about type mismatches.

That's no longer the case in the current version. I've removed the type cast.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1707)*

**Comment by enisoc** [comment]
Reviewed 1 of 1 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[go/vt/binlog/binlogplayer/binlog_player.go, line 396 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4jWXSDpEsXLHU-3UV:-KI50Yhe8hQp7ybvAVye:272386005) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/binlog/binlogplayer/binlog_player.go#L396)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> The maximum backoff duration is 1 second. (I've clarified this in the comment of the Throttler.Throttle() method.)
> 
> Usually, it's 1 second / rate e.g. a rate of 1000 TPS will result into a backoff of 1 ms.
> 
> Given this, I would not add a select for a cancelled context. Let me know if you think otherwise.
</details>
That sounds reasonable. I can wait an extra 1s when trying to stop the binlog player.

Maybe add a comment on the `Sleep()` call here to say we don't bother checking for context cancellation because xyz.

---

*[go/vt/tabletmanager/binlog_players.go, line 217 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4lUD5a3ruj1dfPEWm:-KI51vj9q4pZk6YpRdPL:-1941053869) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/tabletmanager/binlog_players.go#L217)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> My intention was to have something like an assertion here. I overlooked that log.Fatalf() does something else than panic(). Thanks for pointing this out :)
> 
> So panic() would make sense here. However, Iteration() does call recover() and catch any panics. Therefore, it basically renders any panic() into an error and we could use an error in the first place.
> 
> I was also thinking about returning an error but I wanted to have something stronger in the style of an assertion here because I consider it a programming error to call Close() on a non-stopped binlog player.
> 
> Returning an error instead of crashing the process would be better if we assume that our tests do not cover all programming errors. This way we avoid the worst case that a destination master in production could (constantly) crash. For empty (non-serving) destination shards this is not a problem. However, we also want to reshard into existing shards in the future. Given that I'm learning towards returning an error here.
> 
> The fact that there's already a panic handler also suggests that I shouldn't kill the process here.
> 
> Thoughts?
</details>
Is there a reason you don't want `Close()` to imply `Stop()` also, if it hasn't been stopped?

---

*[go/vt/throttler/thread_throttler.go, line 80 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI52oQ14jg0udWQ6taK:-KI52oQ14jg0udWQ6taL:-1400044220) ([raw file](https://github.com/youtube/vitess/blob/4472745429ce4f85955f7878bf7bc13ac225dc2b/go/vt/throttler/thread_throttler.go#L80)):*
> ```Go
> 	//       the returned backoff will always be zero.
> 	// Minimum time between two requests.
> 	requestIntervalNs := 1 * time.Second.Nanoseconds() / maxRate
> ```

What is the multiplication by `1` for?

---

*[go/vt/throttler/throttler.go, line 35 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4mJoU7sww3bTZWbTV:-KI52T61AoxC-qFnNAyA:-958477057) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/throttler/throttler.go#L35)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> With typed constant you mean the explicit cast?
> 
> I had int64(-1) in a previous version because the compiler complained about type mismatches.
> 
> That's no longer the case in the current version. I've removed the type cast.
</details>
Numeric constants can be either typed or untyped. By doing the cast, you declared `MaxRateModuleDisabled` as a constant of type `int64`. If you leave off the type, `MaxRateModuleDisabled` remains an untyped constant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1707)*

**Comment by michael-berlin** [comment]
Review status: 10 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[go/vt/binlog/binlogplayer/binlog_player.go, line 396 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4jWXSDpEsXLHU-3UV:-KI9bVkghLzyy3k4kDZA:-499146009) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/binlog/binlogplayer/binlog_player.go#L396)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> That sounds reasonable. I can wait an extra 1s when trying to stop the binlog player.
> 
> Maybe add a comment on the `Sleep()` call here to say we don't bother checking for context cancellation because xyz.
</details>
Done.

---

*[go/vt/tabletmanager/binlog_players.go, line 217 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI4lUD5a3ruj1dfPEWm:-KI9bqpsj6oGVPT4rjbh:1778616484) ([raw file](https://github.com/youtube/vitess/blob/7c1ffebacfec41eae5edf348022474c222f6f05c/go/vt/tabletmanager/binlog_players.go#L217)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Is there a reason you don't want `Close()` to imply `Stop()` also, if it hasn't been stopped?
</details>
We can probably do that. When Close() is called, it's always directly after a Stop().

On the other hand, we would self-correct a programming error in this case.

Instead of doing this, I've changed Close() to return an error when the binlog player was not stopped. It will be logged at the error level.

---

*[go/vt/throttler/thread_throttler.go, line 80 \[r3\]](https://reviewable.io:443/reviews/youtube/vitess/1707#-KI52oQ14jg0udWQ6taK:-KI9flM_Zjdh98eSxNAs:338579572) ([raw file](https://github.com/youtube/vitess/blob/4472745429ce4f85955f7878bf7bc13ac225dc2b/go/vt/throttler/thread_throttler.go#L80)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> What is the multiplication by `1` for?
</details>
I wanted to be explicit about the fact that we track the current rate only within the current second. (And therefore, the maximum backoff is also limited to 1 second.)

I've put the multiplication into parentheses to make it less confusing:

```go
requestIntervalNs := (1 * time.Second).Nanoseconds() / maxRate
```

In general, I've read somewhere internally as a comment that you should not leave out 1* multiplications e.g. always write ```1 * time.Second``` instead of ```time.Second``` when you mean "1 second". I've followed this in other places of the code as well e.g. on line 63:

```go
		return t.currentSecond.Add(1 * time.Second).Sub(now)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1707)*

**Pull request merged**